src/analyses/base.ml

```diff
@@ -1050,6 +1069,9 @@ struct
      * not include the flag. *)
     let update_one_addr (x, offs) (nst, fl, dep): store =
       let cil_offset = Offs.to_cil_offset offs in
+      let t = match t_override with
```

Comment:
When is this needed?

Reply:
The `return_var ()` points to
```
  let return_varstore = ref dummyFunDec.svar
  let return_varinfo () = !return_varstore
  let return_var () = AD.from_var (return_varinfo ())
  let return_lval (): lval = (Var (return_varinfo ()), NoOffset)
```
which has `void` type.

In set we make sure to cast the argument to the type of the Lval we actually have (since CIL can not insert these casts when writes go through pointers). But here a cast to void would be harmful. To avoid only getting `top` from function invocations, we add this special case.

Without it: `7 tests failed: ["01/04 functions", "01/06 pointers", "01/09 arrays", "01/14 vararg", "01/20 static_var", "22/05 adapted_from_01_09_array", "23/05 adapted_from_01_09_array"]`


https://github.com/goblint/analyzer/blob/ee0c0313cb7091f81049f6b37362da50e1969627/src/analyses/base.ml#L1669-L1690

Reply:
> In set we make sure to cast the argument to the type of the Lval we actually have

I thought we do this before calling `set`.

> Without it: 7 tests failed

But only together with the other changes in this PR?

Reply:
> I thought we do this before calling set.

Yes, but if we write through a pointer, we don't know the  **actual** type of the `Lval` until we get to this point.

> But only together with the other changes in this PR?

Yes.